capstone/src/main/java/com/google/sps/data/RecommenderAlgorithm.java

```diff
@@ -11,49 +13,158 @@
+        Query query = new Query(Room.ROOM_ENTITY_NAME);
+        PreparedQuery results = datastore.prepare(query);
+        EmbeddedEntity entity = user.getEmbeddedTags();
+
+        if (results.countEntities() == 0) {
+            return;
+        }
+
+        entity = Tag.constructUserEmbeddedEntity();
+        for (Entity result : results.asIterable()) {
+            Room room = new Room(result);
+            if (room.getAllFollowers().contains(user.getUserKey())) {
+                for (String tag : room.getAllTags()) {
+                    double frequency = (Double) entity.getProperty(tag);
+                    entity.setProperty(tag, frequency + 1.0);
+                }
             }
         }
+        user.setEmbeddedTags(entity);
+    }
+
+    public static void normalizeEmbeddedEntity(User user) {
+        RecommenderAlgorithm.populateEmbeddedEntity(user);
+        EmbeddedEntity entity = user.getEmbeddedTags();
+        double magnitude = RecommenderAlgorithm.magnitude(entity);
+        for (String tag : Tag.CATEGORY_TAGS) {
+            double value = (Double) entity.getProperty(tag);
+            if (magnitude != 0.0) {
+                entity.setProperty(tag, value/magnitude);
+            }
+        }
+        user.setEmbeddedTags(entity);
+    }
+
+    public static void normalizeAllEmbeddedEntities() {
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Query query = new Query(User.USER_ENTITY_NAME);
+        PreparedQuery results = datastore.prepare(query);
 
-        return userTagMap;
+        for (Entity entity : results.asIterable()) {
+            try{
+                RecommenderAlgorithm.normalizeEmbeddedEntity(DatabaseService.getUser(entity.getKey()));
+            } catch (EntityNotFoundException e) {
+                System.out.println("Entity not found!");
+            }
+        }
     }
+
+    public static Map<Key, Double> getDistanceMaps(User user, EmbeddedEntity entity) {
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Filter tagFilter = new FilterPredicate(User.USER_ID_PROPERTY_KEY, FilterOperator.NOT_EQUAL, user.getId());
+        Query query = new Query(User.USER_ENTITY_NAME).setFilter(tagFilter);
+        PreparedQuery results = datastore.prepare(query);
+
+        Map<Key, Double> userDistancePairs = new TreeMap<>();
+        for (Entity result : results.asIterable()) {
+            Double distance = RecommenderAlgorithm.distanceBetween(entity, (EmbeddedEntity) result.getProperty(User.TAGS_PROPERTY_KEY));
+            userDistancePairs.put(result.getKey(), distance);
+        }
+        return userDistancePairs;
+    }
+
+    public static Double distanceBetween(EmbeddedEntity userTagMap, EmbeddedEntity otherTagMap) {
+        double sum = 0;
+        for (String tag : Tag.CATEGORY_TAGS) {
+            double value = (Double) userTagMap.getProperty(tag) - (Double) otherTagMap.getProperty(tag);
```

Comment:
a variable name like delta helps indicate what the variable means

Reply:
gotcha